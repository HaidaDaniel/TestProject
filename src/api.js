
const PRODUCTS_URL = process.env.REACT_APP_PRODUCTS_URL
const PRODUCTS_URL_BY_ID = process.env.REACT_APP_PRODUCT_BY_ID_URL


export const fetchProducts = async () => {
    try {
        const response = await fetch(PRODUCTS_URL);
        if (!response.ok) {
            throw new Error('Response is not OK');
        }
        const data = await response.json();
        return data;
    } catch (error) {
        console.error('Error in fetchProducts:', error);
        throw error;
    }
};

export const fetchProductById = async (id) => {
    try {
        console.log(PRODUCTS_URL_BY_ID + id)
        const response = await fetch(PRODUCTS_URL_BY_ID + id);
        if (!response.ok) {
            throw new Error('Response is not OK');
        }
        const data = await response.text();
        const jsonData = JSON.parse(data);
        return jsonData.data;
    } catch (error) {
        console.error('Error in fetchProductById:', error);
        throw error;
    }
};