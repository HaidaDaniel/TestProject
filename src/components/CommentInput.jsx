/** @format */
import PropTypes from 'prop-types'
import RatingStars from './RatingStars'
import { useState } from 'react'
import { Form, Button, Modal } from 'react-bootstrap'
import {
  StyledFormGroup,
  StyledFormLabel,
  StyledTextArea,
} from './styled/CommentInputStyles'

function CommentInput({ productId }) {
  const [rating, setRating] = useState(null)
  const [text, setText] = useState('')
  const [showModal, setShowModal] = useState(false)
  const [modalMessage, setModalMessage] = useState('')

  const handleSubmit = async (e) => {
    e.preventDefault()

    const data = {
      productId,
      rating,
      text,
    }

    try {
      const response = await fetch(
        'http://demo5127360.mockable.io/POST/product/1',
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(data),
        }
      )

      if (response.ok) {
        setModalMessage('Review sent successfully')
        setShowModal(true)
      } else {
        setModalMessage('Error sending review')
        setShowModal(true)
      }

      setRating(null)
      setText('')
    } catch (error) {
      console.error('Error sending review:', error)
    }
  }

  const closeModal = () => {
    setShowModal(false)
  }

  return (
    <>
      <Form onSubmit={handleSubmit}>
        <StyledFormGroup>
          <StyledFormLabel>
            Review
            <RatingStars isInput onRatingChange={setRating} />
            <Button type='submit'>Send</Button>
          </StyledFormLabel>
          <StyledTextArea>
            <Form.Control
              as='textarea'
              placeholder='Enter review here ...'
              rows={3}
            />
          </StyledTextArea>
        </StyledFormGroup>
      </Form>
      <Modal show={showModal} onHide={closeModal}>
        <Modal.Header closeButton>
          <Modal.Title>Review Status</Modal.Title>
        </Modal.Header>
        <Modal.Body>{modalMessage}</Modal.Body>
        <Modal.Footer>
          <Button variant='secondary' onClick={closeModal}>
            Close
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  )
}
export default CommentInput

CommentInput.propTypes = {
  productId: PropTypes.string.isRequired,
}
