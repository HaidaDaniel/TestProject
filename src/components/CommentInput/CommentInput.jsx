/** @format */
import RatingStars from '../RatingStars/RatingStars'
import { useState } from 'react'
import { Form, FormGroup, Button } from 'react-bootstrap'

function CommentInput({ productId }) {
  const [rating, setRating] = useState(null)
  const [text, setText] = useState('')

  const handleSubmit = async (e) => {
    e.preventDefault()

    const data = {
      productId,
      rating,
      text,
    }

    await fetch('/api/reviews', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })

    setRating(null)
    setText('')
  }
  return (
    <Form onSubmit={handleSubmit}>
      <RatingStars rating={rating} isInput />
      <FormGroup>
        <label for='text'>Review text</label>
        <textarea
          id='text'
          name='text'
          cols='30'
          rows='10'
          value={text}
          onChange={(e) => setText(e.target.value)}
        />
      </FormGroup>
      <Button type='submit'>Send</Button>
    </Form>
  )
}
export default CommentInput
