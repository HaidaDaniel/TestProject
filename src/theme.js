const theme = {
    colors: {
        primary: '#0d6efd',
        secondary: '#6c757d',
        success: '#28a745',
        error: '#dc3545',
        golden: '#f5a623'
    },
    fonts: {
        body: 'Arial, sans-serif',
        heading: 'Helvetica, sans-serif',
    },

};

export default theme;