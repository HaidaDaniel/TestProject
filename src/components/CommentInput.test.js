import { render, fireEvent, screen } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import '@testing-library/jest-dom/extend-expect';

import theme from '../theme';
import CommentInput from './CommentInput';
import { postReview } from '../api';

jest.mock('../api', () => ({
  postReview: jest.fn(() => Promise.resolve('Review posted successfully')),
}));

describe('CommentInput Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders CommentInput component', () => {
    render(<ThemeProvider theme={theme}><CommentInput productId={1} /></ThemeProvider>);
    const reviewText = screen.getByPlaceholderText('Enter review here ...');
    const sendButton = screen.getByText('Send');

    expect(reviewText).toBeInTheDocument();
    expect(sendButton).toBeInTheDocument();
  });

  it('handles form submission', async () => {
    render(<ThemeProvider theme={theme}><CommentInput productId={1} /></ThemeProvider>);
    const reviewText = screen.getByPlaceholderText('Enter review here ...');
    const sendButton = screen.getByText('Send');

    fireEvent.change(reviewText, { target: { value: 'Test review' } });
    fireEvent.click(sendButton);

    expect(postReview).toHaveBeenCalledWith({
      productId: 1,
      rating: 1,
      text: 'Test review',
    });

    const modal = await screen.findByText('Review Status', { exact: false });
    expect(modal).toBeVisible();

    expect(document.body).toMatchSnapshot();

  });
});