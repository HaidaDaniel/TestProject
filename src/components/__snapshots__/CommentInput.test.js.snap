// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentInput Component handles form submission 1`] = `
<body
  class="modal-open"
  data-rr-ui-modal-open=""
  style="padding-right: 1024px;"
>
  <div>
    <form
      class=""
    >
      <div
        class="sc-ciJnBw kroKEN"
      >
        <label
          class="sc-jdkVqZ fKPHZR form-label"
        >
          Review
          <div
            class="sc-kMbQoj gbwijL"
          >
            <div
              class="sc-cCIIiH dyAKMf"
            >
              <span
                class="sc-ilpitK sc-jIJgYh zgbXG eQxvXi"
              >
                ★
              </span>
            </div>
            <div
              class="sc-cCIIiH dyAKMf"
            >
              <span
                class="sc-ilpitK zgbXG"
              >
                ★
              </span>
            </div>
            <div
              class="sc-cCIIiH dyAKMf"
            >
              <span
                class="sc-ilpitK zgbXG"
              >
                ★
              </span>
            </div>
            <div
              class="sc-cCIIiH dyAKMf"
            >
              <span
                class="sc-ilpitK zgbXG"
              >
                ★
              </span>
            </div>
            <div
              class="sc-cCIIiH dyAKMf"
            >
              <span
                class="sc-ilpitK zgbXG"
              >
                ★
              </span>
            </div>
            <input
              class="sc-gXCJSa kQcHUY"
              max="5"
              min="0"
              step="0.5"
              type="number"
              value="1"
            />
             
            /5
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Send
          </button>
        </label>
        <div
          class="sc-bZCZoC bkWSOt"
        >
          <textarea
            class="form-control"
            placeholder="Enter review here ..."
            rows="3"
          />
        </div>
      </div>
    </form>
  </div>
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block; padding-right: 0px;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Review Status
          </div>
          <button
            aria-label="Close"
            class="btn-close"
            type="button"
          />
        </div>
        <div
          class="modal-body"
        />
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
